htlcswitch/interceptable_switch.go

```diff
@@ -35,75 +35,240 @@ type InterceptableSwitch struct {
 	// htlcSwitch is the underline switch
 	htlcSwitch *Switch
 
-	// fwdInterceptor is the callback that is called for each forward of
-	// an incoming htlc. It should return true if it is interested in handling
-	// it.
-	fwdInterceptor ForwardInterceptor
+	// intercepted is where we stream all intercepted packets coming from
+	// the switch.
+	intercepted    chan []*htlcPacket
+	resolutionChan chan *fwdResolution
+
+	interceptorRegistration chan ForwardInterceptor
+
+	requireInterceptor bool
+
+	interceptor  ForwardInterceptor
+	holdForwards map[channeldb.CircuitKey]*interceptedForward
+
+	quit chan struct{}
 }
 
 // NewInterceptableSwitch returns an instance of InterceptableSwitch.
 func NewInterceptableSwitch(s *Switch) *InterceptableSwitch {
-	return &InterceptableSwitch{htlcSwitch: s}
+	return &InterceptableSwitch{
+		htlcSwitch:              s,
+		intercepted:             make(chan []*htlcPacket),
+		interceptorRegistration: make(chan ForwardInterceptor),
+		holdForwards:            make(map[channeldb.CircuitKey]*interceptedForward),
+		resolutionChan:          make(chan *fwdResolution),
+
+		quit: make(chan struct{}),
+	}
 }
 
 // SetInterceptor sets the ForwardInterceptor to be used.
 func (s *InterceptableSwitch) SetInterceptor(
 	interceptor ForwardInterceptor) {
 
-	s.Lock()
-	defer s.Unlock()
-	s.fwdInterceptor = interceptor
+	select {
+	case s.interceptorRegistration <- interceptor:
```

Comment:
Sync with main loop to prevent race conditions. See also https://github.com/lightningnetwork/lnd/pull/6165#discussion_r791431268